File: quad_logger/scripts/parseQuadBag.m
function [data, varargout] = parseQuadBag(varargin)
% parseQuadBag parse a quad data log file
%   DATA = parseQuadBag uses the default 'quad_log_current' file name to
%   yield a data structure containing select topic data. If this bag does
%   not exist, the user can select the bag via a UI.
%
%   DATA = parseQuadBag(FILENAME) uses the data in a bag with the specified
%   file name, looking in '../bags/'.

% Default empty namespace
namespace = '';

% Default to quad_log_current
if nargin >= 2
    if ~isempty(varargin{2})
        namespace = [varargin{2}, '/'];
    end
end

if nargin == 0
    trialName = 'quad_log_current';
else
    trialName = varargin{1};
end

% Specify the path
filepath = ['../bags/', trialName,'.bag'];
if ~(exist(filepath,'file'))
    disp([filepath, ' does not exist, using UI to specify path']);
    [fileName,pathname] = uigetfile('.bag', 'Select a Bag');
    
    suffix = '.bag';
    trialName = fileName(1:(end-length(suffix)));
    filepath = fullfile(pathname, fileName);
end

% Load the bag
bag = rosbag(filepath);

% Read the state estimate data
stateEstimateData = readMessages(select(bag,'Topic',['/', namespace, 'state/estimate']),'DataFormat','struct');
stateEstimate = struct;
if isempty(stateEstimateData)
    warning('No data on state estimate topic');
else
    stateEstimate.time = cell2mat(cellfun(@(m) double(m.Header.Stamp.Sec) + double(m.Header.Stamp.Nsec)*1E-9, stateEstimateData, 'UniformOutput', 0));
    
    stateEstimate.position = cell2mat(cellfun(@(m) ...
        [m.Body.Pose.Position.X, m.Body.Pose.Position.Y, m.Body.Pose.Position.Z], stateEstimateData, 'UniformOutput', 0));
    stateEstimate.velocity = cell2mat(cellfun(@(m) ...
        [m.Body.Twist.Linear.X, m.Body.Twist.Linear.Y, m.Body.Twist.Linear.Z], stateEstimateData, 'UniformOutput', 0));
    
    stateEstimate.orientationRPY = cell2mat(cellfun(@(m) ...
        fliplr(quat2eul([m.Body.Pose.Orientation.W, m.Body.Pose.Orientation.X, m.Body.Pose.Orientation.Y, m.Body.Pose.Orientation.Z])), stateEstimateData, 'UniformOutput', 0));
    stateEstimate.orientationQuat = cell2mat(cellfun(@(m) ...
        [m.Body.Pose.Orientation.W, m.Body.Pose.Orientation.X, m.Body.Pose.Orientation.Y, m.Body.Pose.Orientation.Z], stateEstimateData, 'UniformOutput', 0));
    stateEstimate.angularVelocity = cell2mat(cellfun(@(m) ...
        [m.Body.Twist.Angular.X, m.Body.Twist.Angular.Y, m.Body.Twist.Angular.Z], stateEstimateData, 'UniformOutput', 0));
    
    stateEstimate.jointPosition = cell2mat(cellfun(@(m) m.Joints.Position.', stateEstimateData, 'UniformOutput', 0));
    stateEstimate.jointVelocity = cell2mat(cellfun(@(m) m.Joints.Velocity.', stateEstimateData, 'UniformOutput', 0));
    stateEstimate.jointEffort = cell2mat(cellfun(@(m) m.Joints.Effort.', stateEstimateData, 'UniformOutput', 0));
end

% Read the ground truth data
stateGroundTruthData = readMessages(select(bag,'Topic',['/', namespace, 'state/ground_truth']),'DataFormat','struct');
stateGroundTruth = struct;
if isempty(stateGroundTruthData)
    warning('No data on ground truth state topic');
else
    
    stateGroundTruth.time = cell2mat(cellfun(@(m) double(m.Header.Stamp.Sec) + double(m.Header.Stamp.Nsec)*1E-9, stateGroundTruthData, 'UniformOutput', 0));
    
    stateGroundTruth.position = cell2mat(cellfun(@(m) ...
        [m.Body.Pose.Position.X, m.Body.Pose.Position.Y, m.Body.Pose.Position.Z], stateGroundTruthData, 'UniformOutput', 0));
    stateGroundTruth.velocity = cell2mat(cellfun(@(m) ...
        [m.Body.Twist.Linear.X, m.Body.Twist.Linear.Y, m.Body.Twist.Linear.Z], stateGroundTruthData, 'UniformOutput', 0));
    
    stateGroundTruth.orientationRPY = cell2mat(cellfun(@(m) ...
        fliplr(quat2eul([m.Body.Pose.Orientation.W, m.Body.Pose.Orientation.X, m.Body.Pose.Orientation.Y, m.Body.Pose.Orientation.Z])), stateGroundTruthData, 'UniformOutput', 0));
    stateGroundTruth.orientationQuat = cell2mat(cellfun(@(m) ...
        [m.Body.Pose.Orientation.W, m.Body.Pose.Orientation.X, m.Body.Pose.Orientation.Y, m.Body.Pose.Orientation.Z], stateGroundTruthData, 'UniformOutput', 0));
    stateGroundTruth.angularVelocity = cell2mat(cellfun(@(m) ...
        [m.Body.Twist.Angular.X, m.Body.Twist.Angular.Y, m.Body.Twist.Angular.Z], stateGroundTruthData, 'UniformOutput', 0));
    
    stateGroundTruth.jointPosition = cell2mat(cellfun(@(m) m.Joints.Position.', stateGroundTruthData, 'UniformOutput', 0));
    stateGroundTruth.jointVelocity = cell2mat(cellfun(@(m) m.Joints.Velocity.', stateGroundTruthData, 'UniformOutput', 0));
    stateGroundTruth.jointEffort = cell2mat(cellfun(@(m) m.Joints.Effort.', stateGroundTruthData, 'UniformOutput', 0));
    
    num_feet = size(stateGroundTruthData{1}.Feet.Feet, 2);
    for i = 1:num_feet
        stateGroundTruth.footPosition{i} = cell2mat(cellfun(@(m) ...
            [m.Feet.Feet(i).Position.X, m.Feet.Feet(i).Position.Y, m.Feet.Feet(i).Position.Z], stateGroundTruthData, 'UniformOutput', 0));
        stateGroundTruth.footVelocity{i} = cell2mat(cellfun(@(m) ...
            [m.Feet.Feet(i).Velocity.X, m.Feet.Feet(i).Velocity.Y, m.Feet.Feet(i).Velocity.Z], stateGroundTruthData, 'UniformOutput', 0));
    end
end

% Read the trajectory data
stateTrajectoryData = readMessages(select(bag,'Topic',['/', namespace, 'state/trajectory']),'DataFormat','struct');
stateTrajectory = struct;
if isempty(stateTrajectoryData)
    warning('No data on trajectory topic');
else
    
    stateTrajectory.time = cell2mat(cellfun(@(m) double(m.Header.Stamp.Sec) + double(m.Header.Stamp.Nsec)*1E-9, stateTrajectoryData, 'UniformOutput', 0));
    
    stateTrajectory.position = cell2mat(cellfun(@(m) ...
        [m.Body.Pose.Position.X, m.Body.Pose.Position.Y, m.Body.Pose.Position.Z], stateTrajectoryData, 'UniformOutput', 0));
    stateTrajectory.velocity = cell2mat(cellfun(@(m) ...
        [m.Body.Twist.Linear.X, m.Body.Twist.Linear.Y, m.Body.Twist.Linear.Z], stateTrajectoryData, 'UniformOutput', 0));
    
    stateTrajectory.orientationRPY = cell2mat(cellfun(@(m) ...
        fliplr(quat2eul([m.Body.Pose.Orientation.W, m.Body.Pose.Orientation.X, m.Body.Pose.Orientation.Y, m.Body.Pose.Orientation.Z])), stateTrajectoryData, 'UniformOutput', 0));
    stateTrajectory.orientationQuat = cell2mat(cellfun(@(m) ...
        [m.Body.Pose.Orientation.W, m.Body.Pose.Orientation.X, m.Body.Pose.Orientation.Y, m.Body.Pose.Orientation.Z], stateTrajectoryData, 'UniformOutput', 0));
    stateTrajectory.angularVelocity = cell2mat(cellfun(@(m) ...
        [m.Body.Twist.Angular.X, m.Body.Twist.Angular.Y, m.Body.Twist.Angular.Z], stateTrajectoryData, 'UniformOutput', 0));
    
    stateTrajectory.jointPosition = cell2mat(cellfun(@(m) m.Joints.Position.', stateTrajectoryData, 'UniformOutput', 0));
    stateTrajectory.jointVelocity = cell2mat(cellfun(@(m) m.Joints.Velocity.', stateTrajectoryData, 'UniformOutput', 0));
    stateTrajectory.jointEffort = cell2mat(cellfun(@(m) m.Joints.Effort.', stateTrajectoryData, 'UniformOutput', 0));
    
    % Omit joint and foot data (not included in reference trajectory)
    stateTrajectory.jointPosition = nan(size(stateTrajectory.jointPosition));
    stateTrajectory.jointVelocity = nan(size(stateTrajectory.jointVelocity));
    stateTrajectory.jointEffort = nan(size(stateTrajectory.jointEffort));
    
    num_feet = size(stateTrajectoryData{1}.Feet.Feet, 2);
    for i = 1:num_feet
        stateTrajectory.footPosition{i} = cell2mat(cellfun(@(m) ...
            [m.Feet.Feet(i).Position.X, m.Feet.Feet(i).Position.Y, m.Feet.Feet(i).Position.Z], stateTrajectoryData, 'UniformOutput', 0));
        stateTrajectory.footVelocity{i} = cell2mat(cellfun(@(m) ...
            [m.Feet.Feet(i).Velocity.X, m.Feet.Feet(i).Velocity.Y, m.Feet.Feet(i).Velocity.Z], stateTrajectoryData, 'UniformOutput', 0));
        
        stateTrajectory.footPosition{i} = nan(size(stateTrajectory.footPosition{i}));
        stateTrajectory.footVelocity{i} = nan(size(stateTrajectory.footVelocity{i}));
    end
end

% Read the control GRFs data
controlGRFsData = readMessages(select(bag,'Topic',['/', namespace, 'control/grfs']),'DataFormat','struct');
controlGRFs = struct;
if isempty(controlGRFsData)
    warning('No data on grf control topic');
else
    
    controlGRFs.time = cell2mat(cellfun(@(m) double(m.Header.Stamp.Sec) + double(m.Header.Stamp.Nsec)*1E-9, controlGRFsData, 'UniformOutput', 0));
    num_feet = 4;
    for i = 1:num_feet
        try
            controlGRFs.vectors{i} = cell2mat(cellfun(@(m) ...
                [m.Vectors(i).X, m.Vectors(i).Y, m.Vectors(i).Z], controlGRFsData, 'UniformOutput', 0));
            controlGRFs.points{i} = cell2mat(cellfun(@(m) ...
                [m.Points(i).X, m.Points(i).Y, m.Points(i).Z], controlGRFsData, 'UniformOutput', 0));
            controlGRFs.contactStates{i} = cell2mat(cellfun(@(m) ...
                [m.ContactStates(i), m.ContactStates(i), m.ContactStates(i)], controlGRFsData, 'UniformOutput', 0));
        catch
            controlGRFs.vectors{i} = cell2mat(cellfun(@(m) ...
                [0,0,0], controlGRFsData, 'UniformOutput', 0));
            controlGRFs.points{i} = cell2mat(cellfun(@(m) ...
                [0,0,0], controlGRFsData, 'UniformOutput', 0));
            controlGRFs.contactStates{i} = cell2mat(cellfun(@(m) ...
                [0,0,0], controlGRFsData, 'UniformOutput', 0));
        end
    end
end

% Read the state GRFs data
stateGRFsData = readMessages(select(bag,'Topic',['/', namespace, 'state/grfs']),'DataFormat','struct');
stateGRFs = struct;
if isempty(stateGRFsData)
    warning('No data on grf state topic');
else
    
    stateGRFs.time = cell2mat(cellfun(@(m) double(m.Header.Stamp.Sec) + double(m.Header.Stamp.Nsec)*1E-9, stateGRFsData, 'UniformOutput', 0));
    num_feet = 4;
    for i = 1:num_feet
        try
            stateGRFs.vectors{i} = cell2mat(cellfun(@(m) ...
                [m.Vectors(i).X, m.Vectors(i).Y, m.Vectors(i).Z], stateGRFsData, 'UniformOutput', 0));
            stateGRFs.points{i} = cell2mat(cellfun(@(m) ...
                [m.Points(i).X, m.Points(i).Y, m.Points(i).Z], stateGRFsData, 'UniformOutput', 0));
            stateGRFs.contactStates{i} = cell2mat(cellfun(@(m) ...
                [m.ContactStates(i), m.ContactStates(i), m.ContactStates(i)], stateGRFsData, 'UniformOutput', 0));
        catch
            stateGRFs.vectors{i} = cell2mat(cellfun(@(m) ...
                [0,0,0], stateGRFsData, 'UniformOutput', 0));
            stateGRFs.points{i} = cell2mat(cellfun(@(m) ...
                [0,0,0], stateGRFsData, 'UniformOutput', 0));
            stateGRFs.contactStates{i} = cell2mat(cellfun(@(m) ...
                [0,0,0], stateGRFsData, 'UniformOutput', 0));
        end
    end
end

% Read the local plan data
localPlanData = readMessages(select(bag,'Topic',['/', namespace, 'local_plan']),'DataFormat','struct');
localPlan = struct;
if isempty(localPlanData)
    warning('No data on local plan topic');
else
    localPlan.time = cell2mat(cellfun(@(m) double(m.StateTimestamp.Sec) + double(m.StateTimestamp.Nsec)*1E-9, localPlanData, 'UniformOutput', 0));
    localPlan.elementTimes = cellfun(@(m) double(m.Diagnostics.ElementTimes'), localPlanData, 'UniformOutput', 0);
    localPlan.solveTime = cell2mat(cellfun(@(m) m.Diagnostics.ComputeTime, localPlanData, 'UniformOutput', 0));
    localPlan.cost = cell2mat(cellfun(@(m) m.Diagnostics.Cost, localPlanData, 'UniformOutput', 0));
    localPlan.iterations = cell2mat(cellfun(@(m) m.Diagnostics.Iterations, localPlanData, 'UniformOutput', 0));
    localPlan.horizonLength = cell2mat(cellfun(@(m) m.Diagnostics.HorizonLength, localPlanData, 'UniformOutput', 0));
    localPlan.complexitySchedule = cellfun(@(m) double(m.Diagnostics.ComplexitySchedule'), localPlanData, 'UniformOutput', 0);
end

% Localize time to the first message
startTime = stateGroundTruth.time(1);
data = struct;

% Update time of existing messages and pack into struct
if ~isempty(fieldnames(stateEstimate))
    stateEstimate.time = stateEstimate.time - startTime;
    data.stateEstimate = stateEstimate;
else
    data.stateEstimate = [];
end

if ~isempty(fieldnames(stateGroundTruth))
    stateGroundTruth.time = stateGroundTruth.time - startTime;
    data.stateGroundTruth = stateGroundTruth;
else
    data.stateGroundTruth = [];
end

if ~isempty(fieldnames(stateTrajectory))
    stateTrajectory.time = stateTrajectory.time - startTime;
    data.stateTrajectory = stateTrajectory;
else
    data.stateTrajectory = [];
end

if ~isempty(fieldnames(controlGRFs))
    controlGRFs.time = controlGRFs.time - startTime;
    data.controlGRFs = controlGRFs;
else
    data.controlGRFs = [];
end

if ~isempty(fieldnames(stateGRFs))
    stateGRFs.time = stateGRFs.time - startTime;
    data.stateGRFs = stateGRFs;
else
    data.stateGRFs = [];
end

if ~isempty(fieldnames(localPlan))
    localPlan.time = localPlan.time - startTime;
    for i = 1:length(localPlan.elementTimes)
        shiftedElementTimes = localPlan.elementTimes{i} + localPlan.time(i);
        localPlan.elementTimes{i} = shiftedElementTimes;
    end
    data.localPlan = localPlan;
else
    data.localPlan = [];
end

% If prompted, return the name of the filename
if (nargout>1)
    varargout{1} = trialName;
end

File: quad_logger/scripts/parse_quad_bag.py
""" MCAP log file parser for quadruped robotics data. """

import os
from collections import defaultdict
from tkinter import Tk, filedialog
import numpy as np
from mcap.reader import make_reader
from mcap_ros2.decoder import DecoderFactory
from scipy.spatial.transform import Rotation as R


def parse_quad_bag(trial_name=None):
    """
    Parse MCAP log file and extract quadruped robotics data.

    Args:
        trial_name (str, optional): Name of the trial log file (without .mcap extension).
            Defaults to "quad_log_current" if None.
        namespace (str, optional): Robot namespace (currently unused). Defaults to "".

    Returns:
        list: A list containing:
            - dict: Parsed data with keys for state_estimate, state_ground_truth,
                   state_trajectory, control_grfs, state_grfs, and local_plan
            - str: The trial name

    Raises:
        FileNotFoundError: If no log file is selected when file dialog is shown.
    """
    def load_mcap(trial_name=None):
        # Default to quad_log_current
        if trial_name is None:
            trial_name = "quad_log_current"

        # Construct expected path
        file_path = f"../bags/{trial_name}.mcap"

        # If file does not exist, open file chooser
        if not os.path.exists(file_path):
            print(f"{file_path} does not exist, choose a file manually...")

            root = Tk()
            root.withdraw()
            file_path = filedialog.askopenfilename(
                title="Select log file",
                filetypes=[("MCAP files", "*.mcap")]
            )
            root.destroy()

            if not file_path:
                raise FileNotFoundError("No log file selected.")

            # Extract trial name from selected file
            trial_name = os.path.basename(file_path).replace(".mcap", "")

        print(f"Loading: {file_path}...")

        # Open the MCAP file for reading
        with open(file_path, "rb") as f:
            reader = make_reader(f, decoder_factories=[DecoderFactory()])

        return reader, file_path, trial_name

    reader, file_path, trial_name = load_mcap()
    print(file_path)

    topic_data = defaultdict(list)
    topic_timestamps = defaultdict(list)

    with open(file_path, "rb") as f:
        reader = make_reader(f, decoder_factories=[DecoderFactory()])
        for _, channel, message, ros_msg in reader.iter_decoded_messages(
            topics=[
                "/robot_1/state/estimate",
                "/robot_1/state/ground_truth",
                "/robot_1/state/trajectory",
                "/robot_1/control/grfs",
                "/robot_1/state/grfs",
                "/robot_1/local_plan"]):
            topic = channel.topic
            timestamp = message.log_time / 1e9  # convert nanosec to sec
            topic_data[topic].append(ros_msg)
            topic_timestamps[topic].append(timestamp)

    # Read state estimate data
    state_estimate_data = topic_data.get("/robot_1/state/estimate")
    state_estimate = {}

    if state_estimate_data is None:
        print('No data on state estimate topic')
    else:
        state_estimate["time"] = np.array(
            topic_timestamps.get("/robot_1/state/estimate"))
        state_estimate["position"] = np.array([[
            m.body.pose.position.x,
            m.body.pose.position.y,
            m.body.pose.position.z] for m in state_estimate_data])
        state_estimate["velocity"] = np.array([[
            m.body.twist.linear.x,
            m.body.twist.linear.y,
            m.body.twist.linear.z] for m in state_estimate_data])
        state_estimate["orientationQuat"] = np.array([[
            m.body.pose.orientation.w,
            m.body.pose.orientation.x,
            m.body.pose.orientation.y,
            m.body.pose.orientation.z] for m in state_estimate_data])
        state_estimate["orientationRPY"] = np.array([
            R.from_quat([m.body.pose.orientation.x,
                         m.body.pose.orientation.y,
                         m.body.pose.orientation.z,
                         m.body.pose.orientation.w]).as_euler("xyz")
            for m in state_estimate_data])
        state_estimate["angularVelocity"] = np.array([[
            m.body.twist.angular.x,
            m.body.twist.angular.y,
            m.body.twist.angular.z] for m in state_estimate_data])
        state_estimate["jointPosition"] = np.array([
            np.array(m.joints.position) for m in state_estimate_data])
        state_estimate["jointVelocity"] = np.array([
            np.array(m.joints.velocity) for m in state_estimate_data])
        state_estimate["jointEffort"] = np.array([
            np.array(m.joints.effort) for m in state_estimate_data])

    # Read ground truth data
    state_ground_truth_data = topic_data.get("/robot_1/state/ground_truth")
    state_ground_truth = {}

    if state_ground_truth_data is None:
        print('No data on ground truth state topic')
    else:
        state_ground_truth["time"] = np.array([
            float(m.header.stamp.sec) + float(m.header.stamp.nanosec) * 1e-9
            for m in state_ground_truth_data])
        state_ground_truth["position"] = np.array([
            [m.body.pose.position.x,
             m.body.pose.position.y,
             m.body.pose.position.z]
            for m in state_ground_truth_data])
        state_ground_truth["velocity"] = np.array([
            [m.body.twist.linear.x,
             m.body.twist.linear.y,
             m.body.twist.linear.z]
            for m in state_ground_truth_data])
        state_ground_truth["orientationQuat"] = np.array([
            [m.body.pose.orientation.w,
             m.body.pose.orientation.x,
             m.body.pose.orientation.y,
             m.body.pose.orientation.z]
            for m in state_ground_truth_data])
        state_ground_truth["orientationRPY"] = np.array([
            R.from_quat([m.body.pose.orientation.x,
                         m.body.pose.orientation.y,
                         m.body.pose.orientation.z,
                         m.body.pose.orientation.w]).as_euler("xyz")
            for m in state_ground_truth_data])
        state_ground_truth["angularVelocity"] = np.array([
            [m.body.twist.angular.x,
             m.body.twist.angular.y,
             m.body.twist.angular.z]
            for m in state_ground_truth_data])
        state_ground_truth["jointPosition"] = np.vstack([
            np.array(m.joints.position)
            for m in state_ground_truth_data])
        state_ground_truth["jointVelocity"] = np.vstack([
            np.array(m.joints.velocity)
            for m in state_ground_truth_data])
        state_ground_truth["jointEffort"] = np.vstack([
            np.array(m.joints.effort)
            for m in state_ground_truth_data])
        num_feet = len(state_ground_truth_data[0].feet.feet)
        state_ground_truth["footPosition"] = []
        state_ground_truth["footVelocity"] = []
        for i in range(num_feet):
            state_ground_truth["footPosition"].append(np.array([
                [m.feet.feet[i].position.x,
                 m.feet.feet[i].position.y,
                 m.feet.feet[i].position.z]
                for m in state_ground_truth_data]))
            state_ground_truth["footVelocity"].append(np.array([
                [m.feet.feet[i].velocity.x,
                 m.feet.feet[i].velocity.y,
                 m.feet.feet[i].velocity.z]
                for m in state_ground_truth_data]))

    # Read trajectory data
    state_trajectory_data = topic_data.get("/robot_1/state/trajectory")
    state_trajectory = {}

    if state_trajectory_data is None:
        print('No data on trajectory topic')
    else:
        state_trajectory["time"] = np.array([
            float(m.header.stamp.sec) + float(m.header.stamp.nanosec) * 1e-9
            for m in state_trajectory_data])
        state_trajectory["position"] = np.array([
            [m.body.pose.position.x,
             m.body.pose.position.y,
             m.body.pose.position.z]
            for m in state_trajectory_data])
        state_trajectory["velocity"] = np.array([
            [m.body.twist.linear.x,
             m.body.twist.linear.y,
             m.body.twist.linear.z]
            for m in state_trajectory_data])
        state_trajectory["orientationQuat"] = np.array([
            [m.body.pose.orientation.w,
             m.body.pose.orientation.x,
             m.body.pose.orientation.y,
             m.body.pose.orientation.z]
            for m in state_trajectory_data])
        state_trajectory["orientationRPY"] = np.array([
            R.from_quat([m.body.pose.orientation.x,
                         m.body.pose.orientation.y,
                         m.body.pose.orientation.z,
                         m.body.pose.orientation.w]).as_euler("xyz")
            for m in state_trajectory_data])
        state_trajectory["angularVelocity"] = np.array([
            [m.body.twist.angular.x,
             m.body.twist.angular.y,
             m.body.twist.angular.z]
            for m in state_trajectory_data])

    # Read control GRF data
    control_grfs_data = topic_data.get("/robot_1/control/grfs")
    control_grfs = {}
    control_grfs["vectors"] = []
    control_grfs["points"] = []
    control_grfs["contactStates"] = []

    if control_grfs_data is None:
        print('No data on grf control topic')
    else:
        control_grfs["time"] = np.array([
            float(m.header.stamp.sec) + float(m.header.stamp.nanosec) * 1e-9
            for m in control_grfs_data])
        num_feet = 4
        for i in range(num_feet):
            try:
                control_grfs["vectors"].append(np.array([
                    [m.vectors[i].x, m.vectors[i].y, m.vectors[i].z]
                    for m in control_grfs_data]))
                control_grfs["points"].append(np.array([
                    [m.points[i].x, m.points[i].y, m.points[i].z]
                    for m in control_grfs_data]))
                control_grfs["contactStates"].append(np.array([
                    [m.contact_states[i],
                     m.contact_states[i],
                     m.contact_states[i]]
                    for m in control_grfs_data]))
            except (IndexError, AttributeError):
                control_grfs["vectors"].append(np.array([
                    [0, 0, 0] for m in control_grfs_data]))
                control_grfs["points"].append(np.array([
                    [0, 0, 0] for m in control_grfs_data]))
                control_grfs["contactStates"].append(np.array([
                    [0, 0, 0] for m in control_grfs_data]))

    # Read state GRF data
    state_grfs_data = topic_data.get("/robot_1/state/grfs")
    state_grfs = {}
    state_grfs["vectors"] = []
    state_grfs["points"] = []
    state_grfs["contactStates"] = []

    if len(state_grfs_data) == 0:
        print('No data on grf state topic')
    else:
        state_grfs["time"] = np.array([
            float(m.header.stamp.sec) + float(m.header.stamp.nanosec) * 1e-9
            for m in state_grfs_data])
        num_feet = 4
        for i in range(num_feet):
            try:
                state_grfs["vectors"].append(np.array([
                    [m.vectors[i].x, m.vectors[i].y, m.vectors[i].z]
                    for m in state_grfs_data]))
                state_grfs["points"].append(np.array([
                    [m.points[i].x, m.points[i].y, m.points[i].z]
                    for m in state_grfs_data]))
                state_grfs["contactStates"].append(np.array([
                    [m.contact_states[i],
                     m.contact_states[i],
                     m.contact_states[i]]
                    for m in state_grfs_data]))
            except (IndexError, AttributeError):
                state_grfs["vectors"].append(np.array([
                    [0, 0, 0] for m in state_grfs_data]))
                state_grfs["points"].append(np.array([
                    [0, 0, 0] for m in state_grfs_data]))
                state_grfs["contactStates"].append(np.array([
                    [0, 0, 0] for m in state_grfs_data]))

    # Read local plan data
    local_plan_data = topic_data.get("/robot_1/local_plan")
    local_plan = {}

    for _, channel, message, ros_msg in reader.iter_decoded_messages(
            topics=["/local_plan"]):
        local_plan_data.append(ros_msg)

    if len(local_plan_data) == 0:
        print('No data on local plan topic')
    else:
        local_plan["time"] = np.array([
            float(m.state_timestamp.sec) +
            float(m.state_timestamp.nanosec) * 1e-9
            for m in local_plan_data])
        local_plan["elementTimes"] = [
            np.array(m.diagnostics.element_times, dtype=float)
            for m in local_plan_data]
        local_plan["solveTime"] = np.array([
            m.diagnostics.compute_time for m in local_plan_data])
        local_plan["cost"] = np.array([
            m.diagnostics.cost for m in local_plan_data])
        local_plan["iterations"] = np.array([
            m.diagnostics.iterations for m in local_plan_data])
        local_plan["horizonLength"] = np.array([
            m.diagnostics.horizon_length for m in local_plan_data])
        local_plan["complexitySchedule"] = [
            np.array(m.diagnostics.complexity_schedule, dtype=float)
            for m in local_plan_data]

    # Localize time to the first message
    start_time = state_ground_truth["time"][0]
    data = {}

    # Update time of existing messages and pack them into the dictionary
    if len(state_estimate.keys()) != 0:
        state_estimate["time"] = state_estimate["time"] - start_time
        data["state_estimate"] = state_estimate
    else:
        data["state_estimate"] = []

    if len(state_ground_truth.keys()) != 0:
        state_ground_truth["time"] = state_ground_truth["time"] - start_time
        data["state_ground_truth"] = state_ground_truth
    else:
        data["state_ground_truth"] = []

    if len(state_trajectory.keys()) != 0:
        state_trajectory["time"] = state_trajectory["time"] - start_time
        data["state_trajectory"] = state_trajectory
    else:
        data["state_trajectory"] = []

    if len(control_grfs.keys()) != 0:
        control_grfs["time"] = control_grfs["time"] - start_time
        data["control_grfs"] = control_grfs
    else:
        data["control_grfs"] = []

    if len(state_grfs.keys()) != 0:
        state_grfs["time"] = state_grfs["time"] - start_time
        data["state_grfs"] = state_grfs
    else:
        data["state_grfs"] = []

    if len(local_plan.keys()) != 0:
        local_plan["time"] = local_plan["time"] - start_time
        for i in range(len(local_plan["elementTimes"])):
            shifted_element_times = (local_plan["elementTimes"][i] +
                                     local_plan["time"][i])
            local_plan["elementTimes"][i] = shifted_element_times
        data["local_plan"] = local_plan
    else:
        data["local_plan"] = []

    return [data, trial_name]
